### ISSUE TITLE
When probably the keras implementation could be released?

### ISSUE THREAD

**Issue by yushuinanrong** [Tetrachrome/subpixel#18]
Hi,

I really appreciate this work, the tensorflow code is nice and clean. Great job！
I'm wondering if there will be a keras version for this project in the future?

Thanks

**Comment by achalshah20** [comment]
Any update on releasing keras version? Thanks!

**Comment by StanislasBertrand** [comment]
If you use keras with tensorflow backend, you can use the given code to create a keras layer that does the PS operation 

Check how a Lambda layer works :
 https://keras.io/layers/core/#lambda

And this is how i coded it, with updates to the reshape and split functions to fit latest tensorflow version :

`
def _phase_shift(I, r):
    # Helper function with main phase shift operation
    input_shape = tf.shape(I)
    bsize = input_shape[0]
    a = input_shape[1]
    b = input_shape[2]
    c = input_shape[3]
    shape = tf.stack([bsize,a,b,r,r])

    input_shape_as_numbers = I.get_shape()
    print(I.get_shape())
    X = tf.reshape(tensor=I, shape=shape)
    X = tf.transpose(X, (0, 1, 2, 4, 3))  # bsize, a, b, 1, 1
    X = tf.split(X, input_shape_as_numbers[1], axis=1)  # a, [bsize, b, r, r]
    X = tf.concat([tf.squeeze(x) for x in X], axis=2)  # bsize, b, a*r, r
    X = tf.split(X,input_shape_as_numbers[2], axis=1)  # b, [bsize, a*r, r]
    X = tf.concat([tf.squeeze(x) for x in X], axis=2)  # bsize, a*r, b*r
    return tf.reshape(X, (bsize, a*r, b*r, 1))

def PS(input_shape, r, color=False):
  # Main OP that you can arbitrarily use in you tensorflow code
	def subpixel_shape(input_shape):
		dims = [input_shape[0],
			input_shape[1] * r,
			input_shape[2] * r,
			int(input_shape[3] / (r ** 2))]
		output_shape = tuple(dims)
		return output_shape
  
	def subpixel(x):
		if color:
			Xc = tf.split(3, 3, X)
			X = tf.concat(3, [_phase_shift(x, r) for x in Xc])
		else:
			x_upsampled = _phase_shift(x, r)
		return x_upsampled

	return Lambda(subpixel, output_shape=subpixel_shape, name='subpixel')
`
Then how it would fit in a model :
`upSampled = PS(x_input.shape, r=3, color=False)(x_input)`

**Comment by StanislasBertrand** [comment]
If you use keras with tensorflow backend, you can use the given code to create a keras layer that does the PS operation

Check how a Lambda layer works :
https://keras.io/layers/core/#lambda

And this is how i coded it, with updates to the reshape and split functions to fit latest tensorflow version :

`def _phase_shift(I, r):
    # Helper function with main phase shift operation
    input_shape = tf.shape(I)
    bsize = input_shape[0]
    a = input_shape[1]
    b = input_shape[2]
    c = input_shape[3]
    shape = tf.stack([bsize,a,b,r,r])

    input_shape_as_numbers = I.get_shape()
    print(I.get_shape())
    X = tf.reshape(tensor=I, shape=shape)
    X = tf.transpose(X, (0, 1, 2, 4, 3))  # bsize, a, b, 1, 1
    X = tf.split(X, input_shape_as_numbers[1], axis=1)  # a, [bsize, b, r, r]
    X = tf.concat([tf.squeeze(x) for x in X], axis=2)  # bsize, b, a*r, r
    X = tf.split(X,input_shape_as_numbers[2], axis=1)  # b, [bsize, a*r, r]
    X = tf.concat([tf.squeeze(x) for x in X], axis=2)  #
    bsize, a*r, b*r
    return tf.reshape(X, (bsize, a*r, b*r, 1))

def PS(input_shape, r, color=False):
  # Main OP that you can arbitrarily use in you tensorflow code
	def subpixel_shape(input_shape):
		dims = [input_shape[0],
			input_shape[1] * r,
			input_shape[2] * r,
			int(input_shape[3] / (r ** 2))]
		output_shape = tuple(dims)
		return output_shape
  
	def subpixel(x):
		if color:
			Xc = tf.split(3, 3, X)
			X = tf.concat(3, [_phase_shift(x, r) for x in Xc])
		else:
			x_upsampled = _phase_shift(x, r)
		return x_upsampled

	return Lambda(subpixel, output_shape=subpixel_shape, name='subpixel')
`
